### PR TITLE
Warn before deleting branch with unmerged/unpushed commits

### LINKED ISSUE #4214: Warn user of unmerged commits before deleting branch
## Description

The user is not warned when deleting a branch which contains commits not yet merged into the master branch.

## Version

* GitHub Desktop: 1.1.0

* Operating system: Windows 10 / Windows 7

## Steps to Reproduce

1. create new branch
2. make one or more commits in the new branch
3. delete the branch

### Expected Behavior

A warning that not all commits have been merged into the master branch
### Actual Behavior

A warning message is displayed stating that the operation cannot be undone. However, the message doesn't mention that there are commits in this branch which have not been merged into master branch

### PR DESCRIPTION
**Pull request by samuelko123**
Closes #4214

## Description
- It addresses the 3 cases [here](https://github.com/desktop/desktop/pull/4217#issuecomment-372131743).
- It shows a `Loader` and runs `git fetch` if remote branch exists, because we want the latest data.
- The `Loader` deliberately blocks the action buttons, because we don't want users to make uninformed decision.

## Discussion
- One potential improvement is to create "Don't show delete branch warning" setting, so user can choose not to run `git fetch`.
- Another potential improvement is the "Unmerged commits" warning message
  - It refers to the number of commits ahead of the default branch, but the branch could've been merged to another branch.
  - It's too expensive to check every commit against every branch.
  - Maybe a better message could help, something like "Warning: `origin/test_branch` has 1 commit that does not exist in `origin/main`". But I feel it's a bit too verbose.

## Test steps
1. Create new branch
2. Try to delete branch, get screenshot No. 1
3. Create changes without committing
4. Try to delete branch, get screenshot No. 1
5. Commit the changes
6. Try to delete branch, get screenshot No. 2
7. Push changes (implicitly create remote branch)
8. Try to delete branch, get screenshot No. 3
9. Create another commit locally
10. Try to delete branch, get screenshot No. 4
11. Merge the remote changes
12. Try to delete branch, get screenshot No. 5

### Screenshots

1.
image

2.
image

3.
image

4.
image

5.
image

6. `git fetch` loading
image

## Release notes

Notes: [Added] Add warning message before deleting a branch with unmerged/unpushed commits

### CODE REVIEW

**Review by iStoreJaber — COMMENTED** [review]
Test

**Review by tidy-dev — CHANGES_REQUESTED** [review]
@samuelko123 I apologize for the long wait on a reply. We briefly discussed this today and we are concerned about the use of fetch. We like the feature, however, fetch, unfortunately, can be an expensive operation for some users and triggers the authentication flow thus could cause authentication requirements for users with shorter authentication cycles. Thus, making the delete branch dialog confirmation really heavy for what we do not view as worth the cost. Because of this, we do not want to move forward with this PR as is. However, if you would like to remove the portion that relies on fetch call. We could take another look. Thank you for making GitHub Desktop better. :) 

Note: If we don't hear back from you soon, we will close it out.

**Comment by samuelko123-bunnings** [comment]
Hi @tidy-dev , thanks for replying.

In this PR, under the "Discussion" section, there is a suggestion of creating a setting for users to decide whether to fetch or not, and I assume you have read and decided it is not a good idea.

Anyway, I have removed the `git fetch` portion from this PR. Hope it can eventually be merged and benefit users to a certain extent.

Thanks for reviewing.

**Comment by samuelko123** [comment]
Hi @tidy-dev  , thanks for replying.

In this PR, under the "Discussion" section, there is a suggestion of creating a setting for users to decide whether to fetch or not, and I assume you have read and decided it is not a good idea.

Anyway, I have removed the `git fetch` portion from this PR. Hope it can eventually be merged and benefit users to a certain extent.

Thanks for reviewing.

File: app/src/lib/remote-parsing.ts [hunk 1]
```diff
@@ -27,15 +27,19 @@
...
   },
   {
     protocol: 'ssh',
-    regex: new RegExp('^git@(.+):(.+)/(.+?)(?:/|.git)?$'),
+    regex: new RegExp('^git@(.+):([^/]+)/([^/]+?)(?:/|.git)?$'),
   },
   {
     protocol: 'ssh',
-    regex: new RegExp('^git:(.+)/(.+)/(.+?)(?:/|.git)?$'),
+    regex: new RegExp('^(?:.+)@(.+.ghe.com):([^/]+)/([^/]+?)(?:/|.git)?$'),
```
[the remark this answers is not in the document]
Comment (samuelko123) [reply app/src/lib/remote-parsing.ts]: I accidentally pushed a merge commit, which caused this warning.
I've force-pushed to remove this merge commit.

**Review by tidy-dev — CHANGES_REQUESTED** [review]
@samuelko123 Sorry for taking so long to circle back. Thank you for removing the fetch logic and you are correct at this time, we do not want to consider adding adding a setting to allow fetching here. 

Some concerns:
- For accessibility, particularly screen reader users, this dialog is an `alertdialog` and any content in it needs to be referenced in the `ariaDescribedBy` of the dialog. Currently, all the warnings added are not included. You could rectify this by placing them all inside a div and adding the `id` of that div to the `ariaDescribedBy` making sure the order of the id's reflects the order they are rendered. Similarly, the 'delete-branch-confirmation-message-remote` id is being removed and we would like to keep it.
- The warnings that say "This branch has x unmerged commits" does not reference to what it is comparing to. We do have users that create branches off of branches other than the default. Suggest that we add make it "This branch has x unmerged commits when compared to `{defaultBranch.name}`" . Similarly for remove ahead but use defaultBranch.upstream.

File: app/src/ui/delete-branch/delete-branch-dialog.tsx [hunk 2]
```diff
@@ -63,15 +120,60 @@ export class DeleteBranch extends React.Component<
...
+  private renderLocalUnmergedWarning(unmergedCommits: number) {
+    if (this.state.existsOnRemote || this.state.localUnmergedCommits <= 0) {
+      return null
+    }
+
+    return (
+      <div>
+        <strong>
+          Warning: This branch has {unmergedCommits} unmerged{' '}
+          {unmergedCommits === 1 ? 'commit' : 'commits'}.
```
Comment (tidy-dev) [inline app/src/ui/delete-branch/delete-branch-dialog.tsx]:
```suggestion
          {unmergedCommits === 1 ? 'commit' : 'commits'} when compared to <Ref>{this.props.defaultBranch.name}</Ref>.
```

^ May cause linting error.. this is just a suggestion to communicate requested change.. needed tested. :)

File: app/src/ui/delete-branch/delete-branch-dialog.tsx [hunk 3]
```diff
@@ -63,15 +120,60 @@ export class DeleteBranch extends React.Component<
...
+  private renderRemoteUnmergedCommitsWarning(unmergedCommits: number) {
+    if (unmergedCommits <= 0) {
+      return null
+    }
+
+    return (
+      <div>
+        <strong>
+          Warning: <Ref>{this.props.branch.upstream}</Ref> has {unmergedCommits}{' '}
+          unmerged {unmergedCommits === 1 ? 'commit' : 'commits'}.
```
Comment (tidy-dev) [inline app/src/ui/delete-branch/delete-branch-dialog.tsx]:
```suggestion
          unmerged {unmergedCommits === 1 ? 'commit' : 'commits'}  when compared to <Ref>{this.props.defaultBranch.upstream}</Ref>.
```

^ May cause linting error.. this is just a suggestion to communicate requested change.. needed tested. :)

File: app/src/ui/delete-branch/delete-branch-dialog.tsx [hunk 4]
```diff
@@ -63,15 +120,60 @@ export class DeleteBranch extends React.Component<
...
-    if (this.props.branch.upstreamRemoteName && this.props.existsOnRemote) {
+    if (this.props.branch.upstreamRemoteName && this.state.existsOnRemote) {
       return (
         <div>
-          <p id="delete-branch-confirmation-message-remote">
-            <strong>
-              The branch also exists on the remote, do you wish to delete it
-              there as well?
-            </strong>
+          <p>
```
Comment (tidy-dev) [inline app/src/ui/delete-branch/delete-branch-dialog.tsx]:
```suggestion
          <p id="delete-branch-confirmation-message-remote">
```
Need to keep for the alert dialog to announce this.

**Comment by samuelko123** [comment]
@tidy-dev Thanks for commenting on this PR. The comments are implemented and tested.
My apology for force-pushing because I accidentally used my work email in VS Code.
(It'll be great if Github Desktop can allow multiple email accounts. 😃 )

**Review by sergiou87 — CHANGES_REQUESTED** [review]
Sorry @samuelko123 it took us this long to get back to you. After our recent accessibility work, I see some extra changes needed in this PR:
- First I'd recommend to update your branch with the latest `development` commit.
- Second, we have removed `br` elements (as you can see in #18444 and #18436)
- Finally, the new added chunks of text should have an `id` in their `div` element, and then that `id` should be added to the `ariaDescribedBy` attribute of the dialog (there is an example of this in the file already)

Thank you for your patience!

**Comment by samuelko123** [comment]
@sergiou87 @tidy-dev 

Thank you for your detailed PR reviews.

I use Github Desktop everyday at work. It’s an awesome software and I wished to help make it better. Unfortunately, my passion and patience on this repo have been eroded over the year.

I fully understand the constraints of resources and prioritisation. There’s nothing personal.

I’ll close this PR for now. Anyone who’s interested could freely use any code within this PR to continue the journey.

Stay safe. 😄

**Pull request closed without merging**